src/wasm/literal.cpp

```diff
@@ -135,6 +141,9 @@ Literal::Literal(const Literal& other) : type(other.type) {
         case HeapType::ext:
           gcData = other.gcData;
           return;
+        case HeapType::exn:
+          new (&exnData) std::shared_ptr<ExnData>(other.exnData);
```

Comment:
So this was the culprit of the segmentation faults. Previously it read:
```c
exnData = other.exnData;
```
but that would execute the destructor of `shared_ptr<ExnData>` on the left-hand-side upon *construction* of this `Literal`, which caused segfaults.

I had copied the approach from `gcData = other.gcData;` 2 lines above. Is that one flawed as well? Should I change it to a `new` as well? Is that other one covered by any test?

Reply:
Hmm, this is at the limits of my C++ expertise, but I believe your new code is right, and the existing gcData code is wrong, and should be fixed. Let me look more into that separately (best to fix it in a separate PR, and I want to find out why tests did not fail - they should have).

Reply:
Ah, there is little mystery why that bad code did not crash... it was never reached, because we handle that case a few lines above :smile: I'll remove it.

https://github.com/WebAssembly/binaryen/blob/4c752b3d5ffecd6dc28fa1c0d15295bc0be60912/src/wasm/literal.cpp#L125-L127